api/overdrive.py

```diff
@@ -1030,25 +1104,50 @@ def get_download_link(self, checkout_response, format_type, error_url):
                 format_list = ", ".join(available_formats)
                 msg = "Could not find specified format %s. Available formats: %s"
                 raise NoAcceptableFormat(
-                    msg % (format_type, ", ".join(available_formats))
+                    msg % (use_format_type, ", ".join(available_formats))
                 )
 
-        return self.extract_download_link(format, error_url)
+        return self.extract_download_link(format, error_url, fetch_manifest)
 
     @classmethod
-    def extract_download_link(cls, format, error_url):
+    def extract_download_link(cls, format, error_url, fetch_manifest=False):
         format_type = format.get('formatType', '(unknown)')
         if not 'linkTemplates' in format:
             raise IOError("No linkTemplates for format %s" % format_type)
         templates = format['linkTemplates']
         if not 'downloadLink' in templates:
             raise IOError("No downloadLink for format %s" % format_type)
-        download_link = templates['downloadLink']['href']
+        download_link_data = templates['downloadLink']
+        if not 'href' in download_link_data:
+            raise IOError("No downloadLink href for format %s" % format_type)
+        download_link = download_link_data['href']
         if download_link:
-            return download_link.replace("{errorpageurl}", error_url)
+            download_link = download_link.replace("{errorpageurl}", error_url)
+            if fetch_manifest:
+                download_link = cls.make_direct_download_link(download_link)
+            return download_link
         else:
             return None
 
+    @classmethod
+    def make_direct_download_link(cls, link):
+        """Convert an Overdrive Read link template to a direct-download link
+        for the manifest.
```

Comment:
I'm confused by the comment. Should this really be "OverDrive Listen link" to "direct download" instead of "OverDrive Read link"?
Or are you using "OverDrive Read" because of the "readauthurl" parameter which appends the additional "contentfile=true"?
(This pertains to #3 in my "conversation".)